File: internal/plugins/jest/rules/prefer_to_be/prefer_to_be_test.go
package prefer_to_be_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/prefer_to_be"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestPreferToBeRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&prefer_to_be.PreferToBeRule,
		[]rule_tester.ValidTestCase{
			// prefer-to-be
			{Code: `expect(null).toBeNull();`},
			{Code: `expect(null).not.toBeNull();`},
			{Code: `expect(null).toBe(1);`},
			{Code: `expect(null).toBe(-1);`},
			{Code: `expect(null).toBe(...1);`},
			{Code: `expect(obj).toStrictEqual([ x, 1 ]);`},
			{Code: `expect(obj).toStrictEqual({ x: 1 });`},
			{Code: `expect(obj).not.toStrictEqual({ x: 1 });`},
			{Code: `expect(value).toMatchSnapshot();`},
			{Code: `expect(catchError()).toStrictEqual({ message: 'oh noes!' })`},
			{Code: `expect("something");`},
			{Code: `expect(token).toStrictEqual(/[abc]+/g);`},
			{Code: `expect(token).toStrictEqual(new RegExp('[abc]+', 'g'));`},
			{Code: "expect(value).toEqual(dedent`my string`);"},
			// prefer-to-be: null
			{Code: `expect(null).not.toEqual();`},
			{Code: `expect(null).toBe();`},
			{Code: `expect(null).toMatchSnapshot();`},
			{Code: `expect("a string").toMatchSnapshot(null);`},
			{Code: `expect("a string").not.toMatchSnapshot();`},
			{Code: `expect(null).toBe`},
			// prefer-to-be: undefined
			{Code: `expect(undefined).toBeUndefined();`},
			{Code: `expect(true).toBeDefined();`},
			{Code: `expect({}).toEqual({});`},
			{Code: `expect(something).toBe()`},
			{Code: `expect(something).toBe(somethingElse)`},
			{Code: `expect(something).toEqual(somethingElse)`},
			{Code: `expect(something).not.toBe(somethingElse)`},
			{Code: `expect(something).not.toEqual(somethingElse)`},
			{Code: `expect(undefined).toBe`},
			// prefer-to-be: NaN
			{Code: `expect(NaN).toBeNaN();`},
			{Code: `expect(true).not.toBeNaN();`},
			{Code: `expect(value).toEqual(null!);`},
			{Code: `expect(value).toEqual(1 satisfies number);`},
			// prefer-to-be: typescript edition
			{Code: `(expect('Model must be bound to an array if the multiple property is true') as any).toHaveBeenTipped()`},
		},
		[]rule_tester.InvalidTestCase{
			// prefer-to-be
			{
				Code:   `expect(value).toEqual("my string");`,
				Output: []string{`expect(value).toBe("my string");`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 15},
				},
			},
			{
				Code:   `expect(value).toStrictEqual("my string");`,
				Output: []string{`expect(value).toBe("my string");`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 15},
				},
			},
			{
				Code:   `expect(value).toStrictEqual(1);`,
				Output: []string{`expect(value).toBe(1);`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 15},
				},
			},
			{
				Code:   `expect(value).toStrictEqual(1,);`,
				Output: []string{`expect(value).toBe(1,);`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 15},
				},
			},
			{
				Code:   `expect(value).toEqual((1));`,
				Output: []string{`expect(value).toBe((1));`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 15},
				},
			},
			{
				Code:   `expect(value).toStrictEqual(-1);`,
				Output: []string{`expect(value).toBe(-1);`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 15},
				},
			},
			{
				Code:   "expect(value).toEqual(`my string`);",
				Output: []string{"expect(value).toBe(`my string`);"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 15},
				},
			},
			{
				Code:   "expect(value)[\"toEqual\"](`my string`);",
				Output: []string{"expect(value)['toBe'](`my string`);"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 15},
				},
			},
			{
				Code:   "expect(value).toStrictEqual(`my ${string}`);",
				Output: []string{"expect(value).toBe(`my ${string}`);"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 15},
				},
			},
			{
				Code:   `expect(loadMessage()).resolves.toStrictEqual("hello world");`,
				Output: []string{`expect(loadMessage()).resolves.toBe("hello world");`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 32},
				},
			},
			{
				Code:   `expect(loadMessage()).resolves["toStrictEqual"]("hello world");`,
				Output: []string{`expect(loadMessage()).resolves['toBe']("hello world");`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 32},
				},
			},
			{
				Code:   `expect(loadMessage())["resolves"].toStrictEqual("hello world");`,
				Output: []string{`expect(loadMessage())["resolves"].toBe("hello world");`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 35},
				},
			},
			{
				Code:   `expect(loadMessage()).resolves.toStrictEqual(false);`,
				Output: []string{`expect(loadMessage()).resolves.toBe(false);`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 32},
				},
			},

			// prefer-to-be: null
			{
				Code:   `expect(null).toBe(null);`,
				Output: []string{`expect(null).toBeNull();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNull", Line: 1, Column: 14},
				},
			},
			{
				Code:   `expect(null).toEqual(null);`,
				Output: []string{`expect(null).toBeNull();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNull", Line: 1, Column: 14},
				},
			},
			{
				Code:   `expect(null).toEqual(null,);`,
				Output: []string{`expect(null).toBeNull();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNull", Line: 1, Column: 14},
				},
			},
			{
				Code:   `expect(null).toStrictEqual(null);`,
				Output: []string{`expect(null).toBeNull();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNull", Line: 1, Column: 14},
				},
			},
			{
				Code:   `expect("a string").not.toBe(null);`,
				Output: []string{`expect("a string").not.toBeNull();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNull", Line: 1, Column: 24},
				},
			},
			{
				Code:   `expect("a string").not["toBe"](null);`,
				Output: []string{`expect("a string").not['toBeNull']();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNull", Line: 1, Column: 24},
				},
			},
			{
				Code:   `expect("a string")["not"]["toBe"](null);`,
				Output: []string{`expect("a string")["not"]['toBeNull']();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNull", Line: 1, Column: 27},
				},
			},
			{
				Code:   `expect("a string").not.toEqual(null);`,
				Output: []string{`expect("a string").not.toBeNull();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNull", Line: 1, Column: 24},
				},
			},
			{
				Code:   `expect("a string").not.toStrictEqual(null);`,
				Output: []string{`expect("a string").not.toBeNull();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNull", Line: 1, Column: 24},
				},
			},

			// prefer-to-be: undefined
			{
				Code:   `expect(undefined).toBe(undefined);`,
				Output: []string{`expect(undefined).toBeUndefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeUndefined", Line: 1, Column: 19},
				},
			},
			{
				Code:   `expect(undefined).toEqual(undefined);`,
				Output: []string{`expect(undefined).toBeUndefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeUndefined", Line: 1, Column: 19},
				},
			},
			{
				Code:   `expect(undefined).toStrictEqual(undefined);`,
				Output: []string{`expect(undefined).toBeUndefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeUndefined", Line: 1, Column: 19},
				},
			},
			{
				Code:   `expect("a string").not.toBe(undefined);`,
				Output: []string{`expect("a string").toBeDefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeDefined", Line: 1, Column: 24},
				},
			},
			{
				Code:   `expect("a string").rejects.not.toBe(undefined);`,
				Output: []string{`expect("a string").rejects.toBeDefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeDefined", Line: 1, Column: 32},
				},
			},
			{
				Code:   `expect("a string").rejects.not["toBe"](undefined);`,
				Output: []string{`expect("a string").rejects['toBeDefined']();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeDefined", Line: 1, Column: 32},
				},
			},
			{
				Code:   `expect("a string")["not"]["toBe"](undefined);`,
				Output: []string{`expect("a string")['toBeDefined']();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeDefined", Line: 1, Column: 27},
				},
			},
			{
				Code:   `expect("a string").not.toEqual(undefined);`,
				Output: []string{`expect("a string").toBeDefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeDefined", Line: 1, Column: 24},
				},
			},
			{
				Code:   `expect("a string").not.toStrictEqual(undefined);`,
				Output: []string{`expect("a string").toBeDefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeDefined", Line: 1, Column: 24},
				},
			},

			// prefer-to-be: NaN
			{
				Code:   `expect(NaN).toBe(NaN);`,
				Output: []string{`expect(NaN).toBeNaN();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNaN", Line: 1, Column: 13},
				},
			},
			{
				Code:   `expect(NaN).toEqual(NaN);`,
				Output: []string{`expect(NaN).toBeNaN();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNaN", Line: 1, Column: 13},
				},
			},
			{
				Code:   `expect(NaN).toStrictEqual(NaN);`,
				Output: []string{`expect(NaN).toBeNaN();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNaN", Line: 1, Column: 13},
				},
			},
			{
				Code:   `expect("a string").not.toBe(NaN);`,
				Output: []string{`expect("a string").not.toBeNaN();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNaN", Line: 1, Column: 24},
				},
			},
			{
				Code:   `expect("a string").rejects.not.toBe(NaN);`,
				Output: []string{`expect("a string").rejects.not.toBeNaN();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNaN", Line: 1, Column: 32},
				},
			},
			{
				Code:   `expect("a string")["rejects"].not.toBe(NaN);`,
				Output: []string{`expect("a string")["rejects"].not.toBeNaN();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNaN", Line: 1, Column: 35},
				},
			},
			{
				Code:   `expect("a string").not.toEqual(NaN);`,
				Output: []string{`expect("a string").not.toBeNaN();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNaN", Line: 1, Column: 24},
				},
			},
			{
				Code:   `expect("a string").not.toStrictEqual(NaN);`,
				Output: []string{`expect("a string").not.toBeNaN();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNaN", Line: 1, Column: 24},
				},
			},

			// prefer-to-be: undefined vs defined
			{
				Code:   `expect(undefined).not.toBeDefined();`,
				Output: []string{`expect(undefined).toBeUndefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeUndefined", Line: 1, Column: 23},
				},
			},
			{
				Code:   `expect(undefined).resolves.not.toBeDefined();`,
				Output: []string{`expect(undefined).resolves.toBeUndefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeUndefined", Line: 1, Column: 32},
				},
			},
			{
				Code:   `expect(undefined).resolves.toBe(undefined);`,
				Output: []string{`expect(undefined).resolves.toBeUndefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeUndefined", Line: 1, Column: 28},
				},
			},
			{
				Code:   `expect("a string").not.toBeUndefined();`,
				Output: []string{`expect("a string").toBeDefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeDefined", Line: 1, Column: 24},
				},
			},
			{
				Code:   `expect("a string").rejects.not.toBeUndefined();`,
				Output: []string{`expect("a string").rejects.toBeDefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeDefined", Line: 1, Column: 32},
				},
			},

			// prefer-to-be: typescript edition
			{
				Code:   `expect(null).toEqual(1 as unknown as string as unknown as any);`,
				Output: []string{`expect(null).toBe(1 as unknown as string as unknown as any);`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 14},
				},
			},
			{
				Code:   `expect(null).toEqual(-1 as unknown as string as unknown as any);`,
				Output: []string{`expect(null).toBe(-1 as unknown as string as unknown as any);`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 14},
				},
			},
			{
				Code:   `expect("a string").not.toStrictEqual("string" as number);`,
				Output: []string{`expect("a string").not.toBe("string" as number);`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBe", Line: 1, Column: 24},
				},
			},
			{
				Code:   `expect(null).toBe(null as unknown as string as unknown as any);`,
				Output: []string{`expect(null).toBeNull();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNull", Line: 1, Column: 14},
				},
			},
			{
				Code:   `expect("a string").not.toEqual(null as number);`,
				Output: []string{`expect("a string").not.toBeNull();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNull", Line: 1, Column: 24},
				},
			},
			{
				Code:   `expect(undefined).toBe(undefined as unknown as string as any);`,
				Output: []string{`expect(undefined).toBeUndefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeUndefined", Line: 1, Column: 19},
				},
			},
			{
				Code:   `expect("a string").toEqual(undefined as number);`,
				Output: []string{`expect("a string").toBeUndefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeUndefined", Line: 1, Column: 20},
				},
			},
			{
				Code:   `expect(value).toEqual((null));`,
				Output: []string{`expect(value).toBeNull();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNull", Line: 1, Column: 15},
				},
			},
			{
				Code:   `expect(value).toEqual((NaN));`,
				Output: []string{`expect(value).toBeNaN();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeNaN", Line: 1, Column: 15},
				},
			},
			{
				Code:   `expect(value).toEqual((undefined));`,
				Output: []string{`expect(value).toBeUndefined();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useToBeUndefined", Line: 1, Column: 15},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/prefer-to-be.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('prefer-to-be', {} as never, {
  valid: [
    { code: 'expect(null).toBeNull();' },
    { code: 'expect(null).not.toBeNull();' },
    { code: 'expect(null).toBe(1);' },
    { code: 'expect(null).toBe(-1);' },
    { code: 'expect(null).toBe(...1);' },
    { code: 'expect(obj).toStrictEqual([ x, 1 ]);' },
    { code: 'expect(obj).toStrictEqual({ x: 1 });' },
    { code: 'expect(obj).not.toStrictEqual({ x: 1 });' },
    { code: 'expect(value).toMatchSnapshot();' },
    { code: "expect(catchError()).toStrictEqual({ message: 'oh noes!' })" },
    { code: 'expect("something");' },
    { code: 'expect(token).toStrictEqual(/[abc]+/g);' },
    { code: "expect(token).toStrictEqual(new RegExp('[abc]+', 'g'));" },
    { code: 'expect(value).toEqual(dedent`my string`);' },

    // null
    { code: 'expect(null).not.toEqual();' },
    { code: 'expect(null).toBe();' },
    { code: 'expect(null).toMatchSnapshot();' },
    { code: 'expect("a string").toMatchSnapshot(null);' },
    { code: 'expect("a string").not.toMatchSnapshot();' },
    { code: 'expect(null).toBe' },

    // undefined
    { code: 'expect(undefined).toBeUndefined();' },
    { code: 'expect(true).toBeDefined();' },

    // NaN
    { code: 'expect(NaN).toBeNaN();' },
    { code: 'expect(true).not.toBeNaN();' },
    { code: 'expect({}).toEqual({});' },
    { code: 'expect(something).toBe()' },
    { code: 'expect(something).toBe(somethingElse)' },
    { code: 'expect(something).toEqual(somethingElse)' },
    { code: 'expect(something).not.toBe(somethingElse)' },
    { code: 'expect(something).not.toEqual(somethingElse)' },
    { code: 'expect(undefined).toBe' },

    // typescript edition
    {
      code: "(expect('Model must be bound to an array if the multiple property is true') as any).toHaveBeenTipped()",
    },
  ],
  invalid: [
    {
      code: 'expect(value).toEqual("my string");',
      output: 'expect(value).toBe("my string");',
      errors: [{ messageId: 'useToBe', column: 15, line: 1 }],
    },
    {
      code: 'expect(value).toStrictEqual("my string");',
      output: 'expect(value).toBe("my string");',
      errors: [{ messageId: 'useToBe', column: 15, line: 1 }],
    },
    {
      code: 'expect(value).toStrictEqual(1);',
      output: 'expect(value).toBe(1);',
      errors: [{ messageId: 'useToBe', column: 15, line: 1 }],
    },
    {
      code: 'expect(value).toStrictEqual(1,);',
      output: 'expect(value).toBe(1,);',
      errors: [{ messageId: 'useToBe', column: 15, line: 1 }],
    },
    {
      code: 'expect(value).toStrictEqual(-1);',
      output: 'expect(value).toBe(-1);',
      errors: [{ messageId: 'useToBe', column: 15, line: 1 }],
    },
    {
      code: 'expect(value).toEqual(`my string`);',
      output: 'expect(value).toBe(`my string`);',
      errors: [{ messageId: 'useToBe', column: 15, line: 1 }],
    },
    {
      code: 'expect(value)["toEqual"](`my string`);',
      output: "expect(value)['toBe'](`my string`);",
      errors: [{ messageId: 'useToBe', column: 15, line: 1 }],
    },
    {
      code: 'expect(value).toStrictEqual(`my ${string}`);',
      output: 'expect(value).toBe(`my ${string}`);',
      errors: [{ messageId: 'useToBe', column: 15, line: 1 }],
    },
    {
      code: 'expect(loadMessage()).resolves.toStrictEqual("hello world");',
      output: 'expect(loadMessage()).resolves.toBe("hello world");',
      errors: [{ messageId: 'useToBe', column: 32, line: 1 }],
    },
    {
      code: 'expect(loadMessage()).resolves["toStrictEqual"]("hello world");',
      output: 'expect(loadMessage()).resolves[\'toBe\']("hello world");',
      errors: [{ messageId: 'useToBe', column: 32, line: 1 }],
    },
    {
      code: 'expect(loadMessage())["resolves"].toStrictEqual("hello world");',
      output: 'expect(loadMessage())["resolves"].toBe("hello world");',
      errors: [{ messageId: 'useToBe', column: 35, line: 1 }],
    },
    {
      code: 'expect(loadMessage()).resolves.toStrictEqual(false);',
      output: 'expect(loadMessage()).resolves.toBe(false);',
      errors: [{ messageId: 'useToBe', column: 32, line: 1 }],
    },

    // null
    {
      code: 'expect(null).toBe(null);',
      output: 'expect(null).toBeNull();',
      errors: [{ messageId: 'useToBeNull', column: 14, line: 1 }],
    },
    {
      code: 'expect(null).toEqual(null);',
      output: 'expect(null).toBeNull();',
      errors: [{ messageId: 'useToBeNull', column: 14, line: 1 }],
    },
    {
      code: 'expect(null).toEqual(null,);',
      output: 'expect(null).toBeNull();',
      errors: [{ messageId: 'useToBeNull', column: 14, line: 1 }],
    },
    {
      code: 'expect(null).toStrictEqual(null);',
      output: 'expect(null).toBeNull();',
      errors: [{ messageId: 'useToBeNull', column: 14, line: 1 }],
    },
    {
      code: 'expect("a string").not.toBe(null);',
      output: 'expect("a string").not.toBeNull();',
      errors: [{ messageId: 'useToBeNull', column: 24, line: 1 }],
    },
    {
      code: 'expect("a string").not["toBe"](null);',
      output: 'expect("a string").not[\'toBeNull\']();',
      errors: [{ messageId: 'useToBeNull', column: 24, line: 1 }],
    },
    {
      code: 'expect("a string")["not"]["toBe"](null);',
      output: 'expect("a string")["not"][\'toBeNull\']();',
      errors: [{ messageId: 'useToBeNull', column: 27, line: 1 }],
    },
    {
      code: 'expect("a string").not.toEqual(null);',
      output: 'expect("a string").not.toBeNull();',
      errors: [{ messageId: 'useToBeNull', column: 24, line: 1 }],
    },
    {
      code: 'expect("a string").not.toStrictEqual(null);',
      output: 'expect("a string").not.toBeNull();',
      errors: [{ messageId: 'useToBeNull', column: 24, line: 1 }],
    },

    // undefined
    {
      code: 'expect(undefined).toBe(undefined);',
      output: 'expect(undefined).toBeUndefined();',
      errors: [{ messageId: 'useToBeUndefined', column: 19, line: 1 }],
    },
    {
      code: 'expect(undefined).toEqual(undefined);',
      output: 'expect(undefined).toBeUndefined();',
      errors: [{ messageId: 'useToBeUndefined', column: 19, line: 1 }],
    },
    {
      code: 'expect(undefined).toStrictEqual(undefined);',
      output: 'expect(undefined).toBeUndefined();',
      errors: [{ messageId: 'useToBeUndefined', column: 19, line: 1 }],
    },
    {
      code: 'expect("a string").not.toBe(undefined);',
      output: 'expect("a string").toBeDefined();',
      errors: [{ messageId: 'useToBeDefined', column: 24, line: 1 }],
    },
    {
      code: 'expect("a string").rejects.not.toBe(undefined);',
      output: 'expect("a string").rejects.toBeDefined();',
      errors: [{ messageId: 'useToBeDefined', column: 32, line: 1 }],
    },
    {
      code: 'expect("a string").rejects.not["toBe"](undefined);',
      output: 'expect("a string").rejects[\'toBeDefined\']();',
      errors: [{ messageId: 'useToBeDefined', column: 32, line: 1 }],
    },
    {
      code: 'expect("a string").not.toEqual(undefined);',
      output: 'expect("a string").toBeDefined();',
      errors: [{ messageId: 'useToBeDefined', column: 24, line: 1 }],
    },
    {
      code: 'expect("a string").not.toStrictEqual(undefined);',
      output: 'expect("a string").toBeDefined();',
      errors: [{ messageId: 'useToBeDefined', column: 24, line: 1 }],
    },

    // NaN
    {
      code: 'expect(NaN).toBe(NaN);',
      output: 'expect(NaN).toBeNaN();',
      errors: [{ messageId: 'useToBeNaN', column: 13, line: 1 }],
    },
    {
      code: 'expect(NaN).toEqual(NaN);',
      output: 'expect(NaN).toBeNaN();',
      errors: [{ messageId: 'useToBeNaN', column: 13, line: 1 }],
    },
    {
      code: 'expect(NaN).toStrictEqual(NaN);',
      output: 'expect(NaN).toBeNaN();',
      errors: [{ messageId: 'useToBeNaN', column: 13, line: 1 }],
    },
    {
      code: 'expect("a string").not.toBe(NaN);',
      output: 'expect("a string").not.toBeNaN();',
      errors: [{ messageId: 'useToBeNaN', column: 24, line: 1 }],
    },
    {
      code: 'expect("a string").rejects.not.toBe(NaN);',
      output: 'expect("a string").rejects.not.toBeNaN();',
      errors: [{ messageId: 'useToBeNaN', column: 32, line: 1 }],
    },
    {
      code: 'expect("a string")["rejects"].not.toBe(NaN);',
      output: 'expect("a string")["rejects"].not.toBeNaN();',
      errors: [{ messageId: 'useToBeNaN', column: 35, line: 1 }],
    },
    {
      code: 'expect("a string").not.toEqual(NaN);',
      output: 'expect("a string").not.toBeNaN();',
      errors: [{ messageId: 'useToBeNaN', column: 24, line: 1 }],
    },
    {
      code: 'expect("a string").not.toStrictEqual(NaN);',
      output: 'expect("a string").not.toBeNaN();',
      errors: [{ messageId: 'useToBeNaN', column: 24, line: 1 }],
    },

    // undefined vs defined
    {
      code: 'expect(undefined).not.toBeDefined();',
      output: 'expect(undefined).toBeUndefined();',
      errors: [{ messageId: 'useToBeUndefined', column: 23, line: 1 }],
    },
    {
      code: 'expect(undefined).resolves.not.toBeDefined();',
      output: 'expect(undefined).resolves.toBeUndefined();',
      errors: [{ messageId: 'useToBeUndefined', column: 32, line: 1 }],
    },
    {
      code: 'expect(undefined).resolves.toBe(undefined);',
      output: 'expect(undefined).resolves.toBeUndefined();',
      errors: [{ messageId: 'useToBeUndefined', column: 28, line: 1 }],
    },
    {
      code: 'expect("a string").not.toBeUndefined();',
      output: 'expect("a string").toBeDefined();',
      errors: [{ messageId: 'useToBeDefined', column: 24, line: 1 }],
    },
    {
      code: 'expect("a string").rejects.not.toBeUndefined();',
      output: 'expect("a string").rejects.toBeDefined();',
      errors: [{ messageId: 'useToBeDefined', column: 32, line: 1 }],
    },

    // typescript edition
    {
      code: 'expect(null).toEqual(1 as unknown as string as unknown as any);',
      output: 'expect(null).toBe(1 as unknown as string as unknown as any);',
      errors: [{ messageId: 'useToBe', column: 14, line: 1 }],
    },
    {
      code: 'expect(null).toEqual(-1 as unknown as string as unknown as any);',
      output: 'expect(null).toBe(-1 as unknown as string as unknown as any);',
      errors: [{ messageId: 'useToBe', column: 14, line: 1 }],
    },
    {
      code: 'expect("a string").not.toStrictEqual("string" as number);',
      output: 'expect("a string").not.toBe("string" as number);',
      errors: [{ messageId: 'useToBe', column: 24, line: 1 }],
    },
    {
      code: 'expect(null).toBe(null as unknown as string as unknown as any);',
      output: 'expect(null).toBeNull();',
      errors: [{ messageId: 'useToBeNull', column: 14, line: 1 }],
    },
    {
      code: 'expect("a string").not.toEqual(null as number);',
      output: 'expect("a string").not.toBeNull();',
      errors: [{ messageId: 'useToBeNull', column: 24, line: 1 }],
    },
    {
      code: 'expect(undefined).toBe(undefined as unknown as string as any);',
      output: 'expect(undefined).toBeUndefined();',
      errors: [{ messageId: 'useToBeUndefined', column: 19, line: 1 }],
    },
    {
      code: 'expect("a string").toEqual(undefined as number);',
      output: 'expect("a string").toBeUndefined();',
      errors: [{ messageId: 'useToBeUndefined', column: 20, line: 1 }],
    },
  ],
});
